velox/functions/prestosql/DateTimeImpl.h

```diff
@@ -60,4 +63,124 @@ FOLLY_ALWAYS_INLINE std::optional<Timestamp> fromUnixtime(double unixtime) {
   auto nanos = unixtime - seconds;
   return Timestamp(seconds, nanos * kNanosecondsInSecond);
 }
+
+namespace {
+enum class DateTimeUnit {
+  kMillisecond,
+  kSecond,
+  kMinute,
+  kHour,
+  kDay,
+  kMonth,
+  kQuarter,
+  kYear
+};
+} // namespace
+
+// Year, quarter or month are not uniformly incremented. Months have different
+// total days, and leap years have more days than the rest. If the new year,
+// quarter or month has less total days than the given one, it will be coerced
+// to use the valid last day of the new month. For example,
+//
+// 2022-01-30 + (1 month) = 2022-02-28
+// 2022-02-28 - (1 month) = 2022-01-28
+//
+// 2022-08-31 + (1 quarter) = 2022-11-30
+// 2022-11-30 - (1 quarter) = 2022-08-30
+//
+// 2020-02-29 + (1 year) = 2021-02-28
+// 2021-02-28 - (1 year) = 2020-02-28
+FOLLY_ALWAYS_INLINE Date
+addToDate(const Date& date, const DateTimeUnit& unit, const int32_t value) {
+  // TODO(gaoge): Handle overflow and underflow with 64-bit representation
+  if (value == 0) {
+    return Date(date.days());
+  }
+
+  const std::chrono::time_point<std::chrono::system_clock, date::days> inDate(
+      date::days(date.days()));
+  std::chrono::time_point<std::chrono::system_clock, date::days> outDate;
+
+  if (unit == DateTimeUnit::kDay) {
+    outDate = inDate + date::days(value);
+  } else {
+    const date::year_month_day calInDate(inDate);
+    date::year_month_day calOutDate;
+
+    if (unit == DateTimeUnit::kMonth) {
+      calOutDate = calInDate + date::months(value);
+    } else if (unit == DateTimeUnit::kQuarter) {
+      calOutDate = calInDate + date::months(3 * value);
+    } else if (unit == DateTimeUnit::kYear) {
+      calOutDate = calInDate + date::years(value);
+    } else {
+      VELOX_UNREACHABLE();
+    }
+
+    if (!calOutDate.ok()) {
```

Comment:
It seems like you are normalizing the day here. I've generally seen code like
constexpr year_month_day normalize(year_month_day ymd){
    ymd += months{0}; // normalizes year and month
    return sys_days{ymd}; // normalizes day
}
Why did you prefer calOutDate = calOutDate.year() / calOutDate.month() / date::last; instead ?